iree/compiler/InputConversion/MHLO/ConvertAndDistributeMHLOToLinalgExt.cpp

```diff
@@ -0,0 +1,163 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/LinalgExt/IR/LinalgExtDialect.h"
+#include "iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree/compiler/InputConversion/MHLO/PassDetail.h"
+#include "iree/compiler/InputConversion/MHLO/Passes.h"
+#include "iree/compiler/InputConversion/MHLO/Rewriters.h"
+#include "mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "mlir-hlo/Dialect/mhlo/transforms/map_lmhlo_to_scalar_op.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+bool isInBodyOfLinalgExtOps(Operation *op) {
+  auto parent_op = op->getParentRegion()->getParentOp();
+  return parent_op->getDialect() ==
+         parent_op->getContext()
+             ->getLoadedDialect<linalg_ext::LinalgExtDialect>();
+}
+
+template <typename OpTy>
+struct LinalgExtRegionHLOOpConversion : public OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      OpTy op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const final {
+    if (!isInBodyOfLinalgExtOps(op)) return failure();
+    if (!op.getResult().getType().template isa<TensorType>()) return failure();
+    if (llvm::all_of(args, [](Value arg) {
+          return arg.getType().template isa<TensorType>();
+        })) {
+      return failure();
+    }
+    Value result = lmhlo::HloOpToStdScalarOp::map<OpTy>(
+        op, getElementTypeOrSelf(op.getType()), args, &rewriter);
+    rewriter.replaceOp(op, result);
+    return success();
+  }
+};
+
+struct LinalgExtRegionReturnOpConversion
+    : public OpConversionPattern<mhlo::ReturnOp> {
+  using OpConversionPattern<mhlo::ReturnOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      mhlo::ReturnOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const final {
+    if (!isInBodyOfLinalgExtOps(op)) return failure();
+    rewriter.replaceOpWithNewOp<linalg_ext::YieldOp>(op, args);
+    return success();
+  }
+};
+
+struct SortOpConversion : public OpConversionPattern<mhlo::SortOp> {
```

Comment:
Not needed for now, but it might be useful to define a base class that does the boiler plate `flow.dispatch.workgroup`s creation and invokes a method implemented in derived class to just convert from `mhlo.*` to `linalg_ext.*` operations. You can use CRTP to avoid virtual dispatches.

Reply:
I think it is a good idea, I made the change in the PR, PTAL.